diff --git a/bazel/dependency_imports.bzl b/bazel/dependency_imports.bzl
index aef33aa103..b7e6fb0d28 100644
--- a/bazel/dependency_imports.bzl
+++ b/bazel/dependency_imports.bzl
@@ -30,7 +30,7 @@ YQ_VERSION = "4.24.4"
 BUF_VERSION = "v1.50.0"
 
 def envoy_dependency_imports(go_version = GO_VERSION, jq_version = JQ_VERSION, yq_version = YQ_VERSION, buf_version = BUF_VERSION):
-    rules_foreign_cc_dependencies()
+    rules_foreign_cc_dependencies(register_default_tools = False, register_built_tools = False)
     go_rules_dependencies()
     go_register_toolchains(go_version)
     if go_version != "host":
