diff --git a/src/docker/local.rs b/src/docker/local.rs
index f7d0507..913b44f 100644
--- a/src/docker/local.rs
+++ b/src/docker/local.rs
@@ -1,3 +1,5 @@
+#![allow(static_mut_refs)]
+
 use std::io;
 use std::path::Path;
 use std::process::{Command, ExitStatus};
diff --git a/src/docker/shared.rs b/src/docker/shared.rs
index 7996e1f..f82db29 100644
--- a/src/docker/shared.rs
+++ b/src/docker/shared.rs
@@ -1,3 +1,5 @@
+#![allow(static_mut_refs)]
+
 use std::io::Write;
 use std::path::{Path, PathBuf};
 use std::process::{Command, ExitStatus, Output};
diff --git a/src/errors.rs b/src/errors.rs
index bcb4d35..7d09934 100644
--- a/src/errors.rs
+++ b/src/errors.rs
@@ -1,3 +1,5 @@
+#![allow(static_mut_refs)]
+
 use crate::docker;
 use crate::temp;
 
diff --git a/src/temp.rs b/src/temp.rs
index f9a180d..0697720 100644
--- a/src/temp.rs
+++ b/src/temp.rs
@@ -1,3 +1,5 @@
+#![allow(static_mut_refs)]
+
 use std::fs;
 use std::path::{Path, PathBuf};
 
