diff --color -urN a/difflib/BUILD.bazel b/difflib/BUILD.bazel
--- a/difflib/BUILD.bazel	1970-01-01 00:00:00.000000000 +0000
+++ b/difflib/BUILD.bazel	2024-12-19 21:26:09.121311218 +0000
@@ -0,0 +1,20 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
+
+go_library(
+    name = "difflib",
+    srcs = ["difflib.go"],
+    importpath = "github.com/pmezard/go-difflib/difflib",
+    visibility = ["//visibility:public"],
+)
+
+alias(
+    name = "go_default_library",
+    actual = ":difflib",
+    visibility = ["//visibility:public"],
+)
+
+go_test(
+    name = "difflib_test",
+    srcs = ["difflib_test.go"],
+    embed = [":difflib"],
+)
