diff --git a/absl/base/BUILD.bazel b/absl/base/BUILD.bazel
index 76122da..ac419f2 100644
--- a/absl/base/BUILD.bazel
+++ b/absl/base/BUILD.bazel
@@ -153,7 +153,7 @@ cc_library(
     copts = ABSL_DEFAULT_COPTS,
     linkopts = select({
         "//absl:windows": [],
-        "//conditions:default": ["-pthread"],
+        "//conditions:default": [],
     }) + ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//visibility:public",
@@ -214,7 +214,7 @@ cc_library(
         "//absl:windows": [
             "-DEFAULTLIB:advapi32.lib",
         ],
-        "//conditions:default": ["-pthread"],
+        "//conditions:default": [],
     }) + ABSL_DEFAULT_LINKOPTS,
     deps = [
         ":atomic_hook",
