diff --git a/shell/private/sh_binary.bzl b/shell/private/sh_binary.bzl
index eec109a..f64848c 100644
--- a/shell/private/sh_binary.bzl
+++ b/shell/private/sh_binary.bzl
@@ -17,6 +17,6 @@
 load(":sh_executable.bzl", "make_sh_executable_rule")
 
 # For doc generation only.
-visibility("public")
+# visibility("public")
 
 sh_binary = make_sh_executable_rule(executable = True)
diff --git a/shell/private/sh_executable.bzl b/shell/private/sh_executable.bzl
index 1306c80..7d06595 100644
--- a/shell/private/sh_executable.bzl
+++ b/shell/private/sh_executable.bzl
@@ -14,7 +14,7 @@
 
 """Common code for sh_binary and sh_test rules."""
 
-visibility(["//shell"])
+# visibility(["//shell"])
 
 _SH_TOOLCHAIN_TYPE = "//shell:toolchain_type"
 
@@ -189,9 +189,9 @@ most build rules</a>.
             "_windows_constraint": attr.label(
                 default = "@platforms//os:windows",
             ),
-        } | extra_attrs,
-        toolchains = [
-            config_common.toolchain_type(_SH_TOOLCHAIN_TYPE, mandatory = False),
-        ],
+        }, # | extra_attrs,
+        # toolchains = [
+        #     config_common.toolchain_type(_SH_TOOLCHAIN_TYPE, mandatory = False),
+        # ],
         **kwargs
     )
diff --git a/shell/sh_binary.bzl b/shell/sh_binary.bzl
index 9e4a30d..8f12326 100644
--- a/shell/sh_binary.bzl
+++ b/shell/sh_binary.bzl
@@ -16,6 +16,6 @@
 
 load("//shell/private:sh_binary.bzl", _sh_binary = "sh_binary")
 
-visibility("public")
+# visibility("public")
 
 sh_binary = getattr(native, "sh_binary", _sh_binary)
