diff --git a/src/lib.rs b/src/lib.rs
index 033e7b46..aa6bcae4 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -122,11 +122,13 @@ const HOST_CONFIG: &str = include_str!(concat!(env!("OUT_DIR"), "/pyo3-build-con
 #[doc(hidden)]
 #[cfg(feature = "resolve-config")]
 fn resolve_cross_compile_config_path() -> Option<PathBuf> {
-    env::var_os("TARGET").map(|target| {
-        let mut path = PathBuf::from(env!("OUT_DIR"));
-        path.push(Path::new(&target));
-        path.push("pyo3-build-config.txt");
-        path
+    env::var_os("TARGET").and_then(|target| {
+        std::env::var("OUT_DIR").ok().map(|out_dir| {
+            let mut path = PathBuf::from(out_dir);
+            path.push(Path::new(&target));
+            path.push("pyo3-build-config.txt");
+            path
+        })
     })
 }
 
