--- a/bazel/dependency_imports.bzl
+++ b/bazel/dependency_imports.bzl
@@ -40,7 +40,7 @@ def envoy_dependency_imports(
         buf_sha = BUF_SHA,
         buf_version = BUF_VERSION):
     compatibility_proxy_repo()
-    rules_foreign_cc_dependencies()
+    rules_foreign_cc_dependencies(register_default_tools = False, register_built_tools = False)
     go_rules_dependencies()
     go_register_toolchains(go_version)
     if go_version != "host":