diff --git a/ossm/vendor/boringssl/include/openssl/target.h b/ossm/vendor/boringssl/include/openssl/target.h
index 2c03389c4..66bb7a44f 100644
--- a/ossm/vendor/boringssl/include/openssl/target.h
+++ b/ossm/vendor/boringssl/include/openssl/target.h
@@ -56,6 +56,8 @@
 #define OPENSSL_32_BIT
 #elif defined(__myriad2__)
 #define OPENSSL_32_BIT
+#elif defined(__s390__) || defined(__s390x__) || defined(__zarch__)
+#define OPENSSL_64_BIT
 #else
 // The list above enumerates the platforms that BoringSSL supports. For these
 // platforms we keep a reasonable bar of not breaking them: automated test
