diff --git a/ossm/vendor/com_google_protobuf/upb/wire/decode.c b/ossm/vendor/com_google_protobuf/upb/wire/decode.c
index 6b6230b57..7e8f89754 100644
--- a/ossm/vendor/com_google_protobuf/upb/wire/decode.c
+++ b/ossm/vendor/com_google_protobuf/upb/wire/decode.c
@@ -359,6 +359,7 @@
       _upb_Decoder_AddEnumValueToUnknown(d, msg, field, &elem);
       continue;
     }
+    _upb_Decoder_MungeInt32(&elem);
     if (_upb_Decoder_Reserve(d, arr, 1)) {
       out = UPB_PTR_AT(upb_Array_MutableDataPtr(arr),
                        arr->UPB_PRIVATE(size) * 4, void);
