diff --git a/ossm/vendor/proxy_wasm_cpp_host/bazel/BUILD b/ossm/vendor/proxy_wasm_cpp_host/bazel/BUILD
index 58b4aed78..4db441139 100644
--- a/ossm/vendor/proxy_wasm_cpp_host/bazel/BUILD
+++ b/ossm/vendor/proxy_wasm_cpp_host/bazel/BUILD
@@ -53,21 +53,8 @@
 )
 
 config_setting(
-    name = "requested_crypto_system",
-    values = {"define": "crypto=system"},
-)
-
-config_setting(
-    name = "linux_s390x",
-    values = {"cpu": "s390x"},
-)
-
-selects.config_setting_group(
     name = "crypto_system",
-    match_any = [
-        ":requested_crypto_system",
-        ":linux_s390x",
-    ],
+    values = {"define": "crypto=system"},
 )
 
 # V8 library with pointer compression disabled using V8's Starlark transition.
diff --git a/ossm/vendor/proxy_wasm_cpp_host/src/exports.cc b/ossm/vendor/proxy_wasm_cpp_host/src/exports.cc
index ea9c2358a..c0ead63d8 100644
--- a/ossm/vendor/proxy_wasm_cpp_host/src/exports.cc
+++ b/ossm/vendor/proxy_wasm_cpp_host/src/exports.cc
@@ -266,7 +266,7 @@
   if (result != WasmResult::Ok) {
     return result;
   }
-  if (!context->wasm()->setDatatype(token_ptr, token)) {
+  if (!context->wasm()->setDatatype(token_ptr, htowasm(token, context->wasmVm()->usesWasmByteOrder()))) {
     return WasmResult::InvalidMemoryAccess;
   }
   return WasmResult::Ok;
@@ -298,7 +298,7 @@
   if (result != WasmResult::Ok) {
     return result;
   }
-  if (!context->wasm()->setDatatype(token_ptr, token)) {
+  if (!context->wasm()->setDatatype(token_ptr, htowasm(token, context->wasmVm()->usesWasmByteOrder()))) {
     return WasmResult::InvalidMemoryAccess;
   }
   return WasmResult::Ok;
@@ -524,12 +524,12 @@
   uint32_t token = 0;
   // NB: try to write the token to verify the memory before starting the async
   // operation.
-  if (!context->wasm()->setDatatype(token_ptr, token)) {
+  if (!context->wasm()->setDatatype(token_ptr, htowasm(token, context->wasmVm()->usesWasmByteOrder()))) {
     return WasmResult::InvalidMemoryAccess;
   }
   auto result =
       context->httpCall(uri.value(), headers, body.value(), trailers, timeout_milliseconds, &token);
-  context->wasm()->setDatatype(token_ptr, token);
+  context->wasm()->setDatatype(token_ptr, htowasm(token, context->wasmVm()->usesWasmByteOrder()));
   return result;
 }
 
@@ -596,7 +596,7 @@
   if (result != WasmResult::Ok) {
     return result;
   }
-  if (!context->wasm()->setDatatype(token_ptr, token)) {
+  if (!context->wasm()->setDatatype(token_ptr, htowasm(token, context->wasmVm()->usesWasmByteOrder()))) {
     return WasmResult::InvalidMemoryAccess;
   }
   return WasmResult::Ok;
@@ -621,7 +621,7 @@
   if (result != WasmResult::Ok) {
     return result;
   }
-  if (!context->wasm()->setDatatype(token_ptr, token)) {
+  if (!context->wasm()->setDatatype(token_ptr, htowasm(token, context->wasmVm()->usesWasmByteOrder()))) {
     return WasmResult::InvalidMemoryAccess;
   }
   return WasmResult::Ok;
