diff --git a/ossm/vendor/toolchains_llvm/toolchain/cc_toolchain_config.bzl b/ossm/vendor/toolchains_llvm/toolchain/cc_toolchain_config.bzl
index 787f37a57..a4db25c1b 100644
--- a/ossm/vendor/toolchains_llvm/toolchain/cc_toolchain_config.bzl
+++ b/ossm/vendor/toolchains_llvm/toolchain/cc_toolchain_config.bzl
@@ -108,6 +108,14 @@ def cc_toolchain_config(
             "clang",
             "glibc_unknown",
         ),
+        "linux-s390x": (
+            "clang-s390x-linux",
+            "s390x",
+            "glibc_unknown",
+            "clang",
+            "clang",
+            "glibc_unknown",
+        ),
         "none-riscv32": (
             "clang-riscv32-none",
             "riscv32",
diff --git a/ossm/vendor/toolchains_llvm/toolchain/internal/common.bzl b/ossm/vendor/toolchains_llvm/toolchain/internal/common.bzl
index 5355d90b7..0f8335f94 100644
--- a/ossm/vendor/toolchains_llvm/toolchain/internal/common.bzl
+++ b/ossm/vendor/toolchains_llvm/toolchain/internal/common.bzl
@@ -17,6 +17,7 @@ SUPPORTED_TARGETS = [
     ("linux", "aarch64"),
     ("linux", "armv7"),
     ("linux", "riscv64"),
+    ("linux", "s390x"),
     ("darwin", "x86_64"),
     ("darwin", "aarch64"),
     ("none", "riscv32"),
diff --git a/ossm/vendor/toolchains_llvm/toolchain/internal/configure.bzl b/ossm/vendor/toolchains_llvm/toolchain/internal/configure.bzl
index 85037739f..9416ab5bf 100644
--- a/ossm/vendor/toolchains_llvm/toolchain/internal/configure.bzl
+++ b/ossm/vendor/toolchains_llvm/toolchain/internal/configure.bzl
@@ -403,6 +403,7 @@ def _cc_toolchain_str(
         "linux-armv7": "armv7-unknown-linux-gnueabihf",
         "linux-x86_64": "x86_64-unknown-linux-gnu",
         "linux-riscv64": "riscv64-unknown-linux-gnu",
+        "linux-s390x": "s390x-unknown-linux-gnu",
         "none-riscv32": "riscv32-unknown-none-elf",
         "none-x86_64": "x86_64-unknown-none",
         "wasm32": "wasm32-unknown-unknown",

