--- glib/glib/valgrind.h
+++ glib/glib/valgrind.h
@@ -643,7 +643,7 @@
                     /* results = r3 */                           \
                     "lgr %0, 3\n\t"                              \
                     : "=d" (_zzq_result)                         \
-                    : "a" (&_zzq_args[0]), "0" (_zzq_default)    \
+                    : "a" (&_zzq_args[0]), "0" ((unsigned long long int)(_zzq_default))    \
                     : "cc", "2", "3", "memory"                   \
                    );                                            \
    _zzq_result;                                                  \
