--- python_configure.bzl
+++ python_configure.bzl
@@ -52,7 +52,7 @@ def _execute(
       the result of repository_ctx.execute(cmdline)
     """
     result = repository_ctx.execute(cmdline)
-    if result.stderr or not (empty_stdout_fine or result.stdout):
+    if result.return_code > 0 or not (empty_stdout_fine or result.stdout):
         _fail("\n".join([
             error_msg.strip() if error_msg else "Repository command failed",
             result.stderr.strip(),
