diff --git a/go/private/platforms.bzl b/go/private/platforms.bzl
index ee6a1c76..ef5319a4 100644
--- a/go/private/platforms.bzl
+++ b/go/private/platforms.bzl
@@ -30,8 +30,7 @@ BAZEL_GOARCH_CONSTRAINTS = {
     "amd64": "@platforms//cpu:x86_64",
     "arm": "@platforms//cpu:armv7",
     "arm64": "@platforms//cpu:aarch64",
-    "ppc64": "@platforms//cpu:ppc",
-    "ppc64le": "@platforms//cpu:ppc64le",
+    "ppc64le": "@platforms//cpu:ppc",
     "s390x": "@platforms//cpu:s390x",
 }
 
diff --git a/go/private/rules/binary.bzl b/go/private/rules/binary.bzl
index 40a17f4d..2741ad71 100644
--- a/go/private/rules/binary.bzl
+++ b/go/private/rules/binary.bzl
@@ -513,8 +513,6 @@ exit /b %GO_EXIT_CODE%
             content = cmd,
         )
         ctx.actions.run(
-            executable = bat,
-            inputs = depset(
-                ctx.files.srcs + [sdk.go],
-                transitive = [sdk.headers, sdk.srcs, sdk.tools],
-            ),
+            executable = "cmd.exe",
+            arguments = ["/S", "/C", bat.path.replace("/", "\\")],
+            inputs = sdk.headers + sdk.tools + sdk.srcs + ctx.files.srcs + [sdk.go, bat],
             outputs = [out, gotmp],
             mnemonic = "GoToolchainBinaryBuild",
         )
